invenio_records_resources/services/records/queryparser/query.py

```diff
@@ -61,19 +61,55 @@ class SearchOptions:
             )
     """
 
-    def __init__(self, identity=None, extra_params=None, tree_transformer_factory=None):
+    def __init__(self, identity=None, extra_params=None, tree_transformer_cls=None):
         """Initialise the parser."""
         self.identity = identity
+        self.tree_transformer_cls = tree_transformer_cls
         self.extra_params = extra_params or {}
-        self.tree_transformer_factory = tree_transformer_factory
+        # the pop or {} is needed due to extra_params being passed from the factory
+        # it is possible that e.g. allow_list=None and then it will fail to set()
+        self.mapping = self.extra_params.pop("mapping", None) or {}
+        self.allow_list = self._calculate_allow_list(
+            self.extra_params.pop("allow_list", None), self.mapping
+        )
+        # process fields list
+        self.fields = self._calculate_fields(
+            # get instead of pop since fields could desired for
+            # both querystring and multi_match
+            self.extra_params.get("fields", None),
+            self.allow_list,
+        )
+
+    def _calculate_allow_list(self, allow_list, mapping):
```

Comment:
does it make sense to have both method `static`, avoiding any temptation of using `self`?